### PR TITLE
Add MAX32690

### PR DESCRIPTION
**Pull request by ozersa**
The [MAX32690 ](https://www.analog.com/en/products/max32690.html) microcontroller (MCU) is an advanced system-on-chip (SoC) featuring an Arm® Cortex®-M4F and RISC-V CPU, large Flash and SRAM memories, and the latest generation Bluetooth® 5 Low Energy (LE) radio.

This PR add MAX32690 in supported zephyr SoCs by providing basic functionality
- Update manifest file
- Add clock control
- Add gpio driver
- Add uart drivers
- Add [MAX32690 ](https://www.analog.com/en/products/max32690.html) SoC support
- Add [MAX32690EVKIT ](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/max32690evkit.html) support

Tested with:
[blinky ](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/basic/blinky) example
image

[button](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/basic/button) example
image

[gpio_basic_api ](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/gpio/gpio_basic_api) test
image

[uart_basic_api ](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/uart/uart_basic_api) test
image

### CODE REVIEW

**Comment by MaureenHelm** [comment]
The compliance checks and doc builds are failing due to:
```
get_maintainer.MaintainersError: MAINTAINERS.yml: glob pattern 'modules/hal_adi/' in 'files' in area 'West project: hal_adi' does not match any files
```

**Comment by ozersa** [comment]
> The compliance checks and doc builds are failing due to:
> 
> ```
> get_maintainer.MaintainersError: MAINTAINERS.yml: glob pattern 'modules/hal_adi/' in 'files' in area 'West project: hal_adi' does not match any files
> ```

Pushed https://github.com/zephyrproject-rtos/zephyr/pull/72291/commits/c7b9c8d9d9276d8e9fcc8735f87d065896b8889f commit to fix it.

File: drivers/gpio/gpio_max32.c [hunk 1]
```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright (c) 2023-2024 Analog Devices, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <pinctrl_soc.h>
```
Comment (gmarull) [inline drivers/gpio/gpio_max32.c:7]: please do not include this header directly, it's meant to be internal.

File: drivers/pinctrl/CMakeLists.txt [hunk 2]
```diff
@@ -37,5 +37,6 @@ zephyr_library_sources_ifdef(CONFIG_PINCTRL_QUICKLOGIC_EOS_S3 pinctrl_eos_s3.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_MCI_IO_MUX pinctrl_mci_io_mux.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_ENE_KB1200 pinctrl_ene_kb1200.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_IMX_SCU pinctrl_imx_scu.c)
+zephyr_library_sources_ifdef(CONFIG_PINCTRL_MAX32 pinctrl_max32.c)
```
Comment (gmarull) [inline drivers/pinctrl/CMakeLists.txt:40]: please, split pinctrl driver into its own commit

File: drivers/gpio/gpio_max32.c [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+#include <gpio.h>
+
+#define DT_DRV_COMPAT adi_max32_gpio
+
+LOG_MODULE_REGISTER(gpio_max32, CONFIG_GPIO_LOG_LEVEL);
+
+struct max32_gpio_config {
+	struct gpio_driver_config common;
+	mxc_gpio_regs_t *regs;
+	const struct pinctrl_dev_config *pctrl;
```
Comment (gmarull) [inline drivers/gpio/gpio_max32.c:24]: where is this used?

File: drivers/pinctrl/pinctrl_max32.c [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "zephyr/devicetree.h"
+#include "zephyr/sys/util_macro.h"
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/max32-pinctrl.h>
+#include <zephyr/drivers/gpio.h>
+#include <gpio/gpio_max32.h>
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_max32.c:12]: please, include headers correctly:

use `<>`, sort by stdlibc, zephyr, 3rd party

File: dts/bindings/pinctrl/adi,max32-pinctrl.yaml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        This macro is available here:
+          -include/zephyr/dt-bindings/pinctrl/max32-pinctrl.h
+        Some examples of macro usage:
+        P0.9 set as alernate function 1
+        {
+          pinmux = <MAX32_PINMUX(0, 9, MAX32_AF1)>;
+        };
+        P0.9 set as alernate function 2
+        {
+          pinmux = <MAX32_PINMUX(0, 9, MAX32_AF2)>;
```
Comment (gmarull) [inline dts/bindings/pinctrl/adi,max32-pinctrl.yaml:67]: `MAX32_` prefix in AF feels redundant, see STM32/GD32 approach.

File: dts/bindings/serial/adi,max32-uart.yaml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      Clock source to be used by the UART peripheral. The following options
+      are available:
+      - 0: "ADI_MAX32_PRPH_CLK_SRC_PCLK" Peripheral clock
+      - 1: "ADI_MAX32_PRPH_CLK_SRC_EXTCLK" External Clock
+      - 2: "ADI_MAX32_PRPH_CLK_SRC_IBRO" Internal Baud Rate Oscillator
+      - 3: "ADI_MAX32_PRPH_CLK_SRC_ERFO" External Radio Frequency Oscillator
+      - 4: "ADI_MAX32_PRPH_CLK_SRC_ERTCO" External Real-Time Clock Oscillator
+      - 5: "ADI_MAX32_PRPH_CLK_SRC_INRO" Internal Ring Oscillator
+      The target device might not support all option please take a look on
+      target device user guide
```
Comment (gmarull) [inline dts/bindings/serial/adi,max32-uart.yaml:36]: provide a header with defines for human readable sources?

File: drivers/serial/Kconfig.max32 [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+# Copyright (c) 2023-2024 Analog Devices, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+config UART_MAX32
+	bool "MAX32 MCU serial driver"
+	default y
+	depends on DT_HAS_ADI_MAX32_UART_ENABLED
+	select SERIAL_HAS_DRIVER
+	select SERIAL_SUPPORT_INTERRUPT
```
Comment (gmarull) [inline drivers/serial/Kconfig.max32:11]: please select PINCTRL where used.

File: soc/adi/max32/common/pinctrl_soc.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+#include <zephyr/types.h>
+
+#include <zephyr/dt-bindings/pinctrl/max32-pinctrl.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct pinctrl_soc_pin {
+	uint32_t pinmux;
```
Comment (gmarull) [inline soc/adi/max32/common/pinctrl_soc.h:20]: wrong commit, please, place all pinctrl stuff in 1 commit

File: soc/adi/max32/soc.c [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <wrap_max32_sys.h>
+
+#if defined(CONFIG_MAX32_ON_ENTER_CPU_IDLE_HOOK)
+bool z_arm_on_enter_cpu_idle(void)
+{
+	/* Returning false prevent device goes to sleep mode */
+	return false;
+}
+#endif
```
Comment (gmarull) [inline soc/adi/max32/soc.c:23]: is this because of a bug?

File: soc/adi/max32/soc.h [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+#ifndef _MAX32_SOC_H_
+#define _MAX32_SOC_H_
+
+#include <wrap_max32xxx.h>
+
+#ifndef _ASMLANGUAGE
+
+#endif /* !_ASMLANGUAGE */
```
Comment (gmarull) [inline soc/adi/max32/soc.h:19]: ?

File: boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+# Enable HEX output, requires for west flash
+CONFIG_BUILD_OUTPUT_HEX=y
+
+# Enable GPIO
+CONFIG_GPIO=y
+
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=120000000
+
+CONFIG_PINCTRL=y
```
Comment (gmarull) [inline boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig:12]: please, make sure drivers using pinctrl select this.

File: boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2023-2024 Analog Devices, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+# Enable HEX output, requires for west flash
+CONFIG_BUILD_OUTPUT_HEX=y
```
Comment (gmarull) [inline boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig:5]: should be a soc level selection

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_max32.c (hunk 1 above)
Comment (ozersa) [reply drivers/gpio/gpio_max32.c]: Thanks, removed.

File: drivers/pinctrl/CMakeLists.txt (hunk 2 above)
Comment (ozersa) [reply drivers/pinctrl/CMakeLists.txt:40]: Thanks, done.

File: drivers/pinctrl/pinctrl_max32.c (hunk 4 above)
Comment (ozersa) [reply drivers/pinctrl/pinctrl_max32.c]: Thanks, done.

File: dts/bindings/pinctrl/adi,max32-pinctrl.yaml (hunk 5 above)
Comment (ozersa) [reply dts/bindings/pinctrl/adi,max32-pinctrl.yaml]: Thanks, redundant prefix removed from comment

File: drivers/serial/Kconfig.max32 (hunk 7 above)
Comment (ozersa) [reply drivers/serial/Kconfig.max32:11]: Thanks, added.

File: soc/adi/max32/common/pinctrl_soc.h (hunk 8 above)
Comment (ozersa) [reply soc/adi/max32/common/pinctrl_soc.h:20]: Thanks, done.

File: soc/adi/max32/soc.c (hunk 9 above)
Comment (ozersa) [reply soc/adi/max32/soc.c:23]: This is hardware limitation, WFI call switch device to the sleep mode, and in sleep mode debugger connection lost. so that we added this feature to control idle state behaviour.

File: soc/adi/max32/soc.h (hunk 10 above)
Comment (ozersa) [reply soc/adi/max32/soc.h]: Redundant, removed, thanks.

File: boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig (hunk 12 above)
Comment (ozersa) [reply boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig:5]: Our debugger support Drag-n-Drop programming too, it requires .bin file.
So that on default generating .bin file simplify things.

File: dts/bindings/serial/adi,max32-uart.yaml (hunk 6 above)
Comment (ozersa) [reply dts/bindings/serial/adi,max32-uart.yaml:36]: Thanks,
We have already added [adi_max32_clock.h](https://github.com/zephyrproject-rtos/zephyr/pull/72291/commits/a8ebbccfbf1b67a8bb2e08033e60ccbc30d96a31#diff-6315da564c6dbec63104a705a7164b256e787db64ebe1d24d031053c303de6b3) file which define all possible clock option, and while using we pass this macro to the devicetree node. see: [max32xxx.dtsi#L135](https://github.com/analogdevicesinc/zephyr/blob/add-MAX32690/dts/arm/adi/max32/max32xxx.dtsi#L135)  and [max32690evkit_max32690_m4.dts#L80](https://github.com/analogdevicesinc/zephyr/blob/add-MAX32690/boards/adi/max32690evkit/max32690evkit_max32690_m4.dts#L80)
Multiple peripheral takes clock-source parameters on MAX32 MCUs. Like UART, WDT, Timer... We have already implemented driver for them and other MAX32 partnumbers, all them rely on this use case. so that I would like to keep existing method to eliminate possible side affect.

**Comment by ozersa** [comment]
@gmarull thanks for your review, it was helpful. please revisit when you get a chance.

File: west.yml [hunk 13]
```diff
@@ -136,6 +138,12 @@ manifest:
       path: modules/fs/fatfs
       groups:
         - fs
+    - name: hal_adi
+      remote: analogdevicesinc
```
Comment (MaureenHelm) [inline west.yml:142]: Can remove this now that https://github.com/zephyrproject-rtos/hal_adi  has been created

File: modules/hal_adi/Kconfig [hunk 14]
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2024 Analog Devices, Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# Place holder if/when Kconfig is needed
```
Comment (MaureenHelm) [inline modules/hal_adi/Kconfig:7]: > Also this file requires to github action works

Which action was failing without this file?

File: MAINTAINERS.yml [hunk 15]
```diff
@@ -4190,6 +4190,22 @@ West:
...
+  status: maintained
+  maintainers:
+    - MaureenHelm
+  collaborators:
+    - ozersa
+    - sihyung-maxim
+    - Jake-Carter
+    - lorne-maxim
+    - galak
+    - microbuilder
```
Comment (MaureenHelm) [inline MAINTAINERS.yml:4203]: Remove

Comment (ozersa) [reply MAINTAINERS.yml]: Removed.

File: modules/hal_adi/Kconfig (hunk 14 above)
Comment (ozersa) [reply modules/hal_adi/Kconfig:7]: It was [doc-build.yml](https://github.com/zephyrproject-rtos/zephyr/actions/runs/8940573961/job/24559300517#step:9:101).
But I believe you are asking me to update commit to mention it.

Thanks commit message updated.

Comment (MaureenHelm) [reply modules/hal_adi/Kconfig:7]: I think instead of introducing the placeholder file you could set `files:` to be empty in `MAINTAINERS.yml`. e.g.,
```yml
"West project: hal_adi":
  status: maintained
  maintainers:
    - MaureenHelm
  collaborators:
    - ozersa
  files: []
  labels:
    - "platform: ADI"
 ```

Comment (ozersa) [reply modules/hal_adi/Kconfig]: Thanks, both of them are fine.
Updates as your suggestion.

**Review by MaureenHelm — DISMISSED** [review]

File: boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig (hunk 12 above)
Comment (gmarull) [reply boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig:5]: This should not be a board level config, but a soc one.

File: boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+
+# Enable GPIO
+CONFIG_GPIO=y
+
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=120000000
```
Comment (gmarull) [inline boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig:10]: why not pull this from DT? See e.g. GD32

File: drivers/gpio/gpio_max32.h [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_GPIO_GPIO_MAX32_H_
+#define ZEPHYR_DRIVERS_GPIO_GPIO_MAX32_H_
+
+#include <zephyr/device.h>
+
+int gpio_max32_config_pinmux(const struct device *dev, int pin, int afx, int pincfg);
```
Comment (gmarull) [inline drivers/gpio/gpio_max32.h:12]: what's going on here?

File: dts/arm/adi/max32/max32690-pinctrl.dtsi [hunk 18]
```diff
@@ -0,0 +1,771 @@
...
+	soc {
+		pinctrl: pin-controller@40008000 {
+
+			/omit-if-no-ref/ spixr_sdio0_p0_1: spixr_sdio0_p0_1 {
+				pinmux = <MAX32_PINMUX(0, 1, AF1)>;
+			};
+
+			/omit-if-no-ref/ spixf_sdio0_p0_1: spixf_sdio0_p0_1 {
+				pinmux = <MAX32_PINMUX(0, 1, AF2)>;
+			};
```
Comment (gmarull) [inline dts/arm/adi/max32/max32690-pinctrl.dtsi:19]: suggestion: pinctrl grouping approach is better, this representation (even if used by some vendors) is pretty dumb.

File: soc/adi/max32/soc.c (hunk 9 above)
Comment (gmarull) [reply soc/adi/max32/soc.c:23]: so are you saying this device can't go to idle? is it busy-looping when going to idle thread?

File: boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig (hunk 12 above)
Comment (ozersa) [reply boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig]: Thanks, moved in [soc/adi/max32/KConfig](https://github.com/zephyrproject-rtos/zephyr/blob/2891de7d6bae8ff7dd0b85b61563bd019bb7e6d3/soc/adi/max32/Kconfig#L12) file.

File: boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig (hunk 16 above)
Comment (ozersa) [reply boards/adi/max32690evkit/max32690evkit_max32690_m4_defconfig]: Because i see most vendors applied this method.
Thanks for the reference, apply it. see: [Kconfig.defconfig.max32690](https://github.com/zephyrproject-rtos/zephyr/blob/2891de7d6bae8ff7dd0b85b61563bd019bb7e6d3/soc/adi/max32/Kconfig.defconfig.max32690#L8)

File: dts/arm/adi/max32/max32690-pinctrl.dtsi (hunk 18 above)
Comment (ozersa) [reply dts/arm/adi/max32/max32690-pinctrl.dtsi:19]: Thanks for suggestion, noted, will apply it when get time. for now please process with existing method.

**Comment by ozersa** [comment]
Rebased onto upstream/main after getting below github action error.
https://github.com/zephyrproject-rtos/zephyr/actions/runs/9187538992/job/25265429018
image

File: soc/adi/max32/soc.c (hunk 9 above)
Comment (ozersa) [reply soc/adi/max32/soc.c:23]: On default yes,
It will go if CONFIG_MAX32_ON_ENTER_CPU_IDLE_HOOK=n

**Comment by MaureenHelm** [comment]
> Rebased onto upstream/main after getting below github action error. [zephyrproject-rtos/zephyr/actions/runs/9187538992/job/25265429018](https://github.com/zephyrproject-rtos/zephyr/actions/runs/9187538992/job/25265429018) image

GitHub is having issues today. I retriggered the jobs.

File: drivers/gpio/gpio_max32.h (hunk 17 above)
Comment (gmarull) [reply drivers/gpio/gpio_max32.h:12]: please, make pinctrl driver self-contained, IIRC only ST has something similar for historical reasons.

Comment (ozersa) [reply drivers/gpio/gpio_max32.h]: Thanks, it is done. please see [pinctrl_max32.c](https://github.com/zephyrproject-rtos/zephyr/blob/23a978e6b17827e4655b4a35d006346c00abe57c/drivers/pinctrl/pinctrl_max32.c)

**Review by MaureenHelm — COMMENTED** [review]
Rebase to latest main to pick up doc build fixes that were merged yesterday.

**Comment by ozersa** [comment]
> Rebase to latest main to pick up doc build fixes that were merged yesterday.

Done.

**Review by MaureenHelm — APPROVED** [review]

**Comment by ozersa** [comment]
@gmarull please revisit when you get a chance.

File: soc/adi/max32/Kconfig [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+	select BUILD_OUTPUT_HEX
+
+config SOC_MAX32690
+	select CPU_CORTEX_M4
+
+if SOC_FAMILY_MAX32
+
+config MAX32_ON_ENTER_CPU_IDLE_HOOK
+	bool "CPU idle hook enable"
+	default y
```
Comment (gmarull) [inline soc/adi/max32/Kconfig:21]: last question: why is this default y? are you preventing sleep by default? maybe default y if DEBUG or something similar would make more sense

Comment (ozersa) [reply soc/adi/max32/Kconfig:21]: Thanks for suggestion,
This parameters requires to access device incase of it goes to sleep mode immediately after power up. So that it requires both for debugging and FW update.
We are working on our OpenOCD fork, trying to handle this case. We are trying to add feature that allow to access MCU even it goes to sleep mode. It almost done. After we release new version of ADI openocd distro we will revisit MAX32_ON_ENTER_CPU_IDLE_HOOK flag. 
So that for new keeping it as it is would be better for MAX32 MCUs.

**Review by gmarull — APPROVED** [review]
